toolchain/lex/lex.cpp

```diff
@@ -1179,17 +1175,43 @@ auto Lexer::LexWordAsTypeLiteralToken(llvm::StringRef word, int32_t byte_offset)
       return LexResult::NoMatch();
   };
 
+  // No leading zeros allowed.
+  if ('1' > word[1] || word[1] > '9') {
+    return LexResult::NoMatch();
+  }
+
   llvm::StringRef suffix = word.substr(1);
-  if (!CanLexInt(emitter_, suffix)) {
+  constexpr ssize_t DigitLimit = std::numeric_limits<int64_t>::digits10;
```

Comment:
Can you test this with a boundary value, such as `9` repeated for the number of digits? i.e., something that fits within the digit count for `ssize_t`, but will overflow a `ssize_t` if lexed.

Reply:
Good idea, added a test that adds 9s successively until it is rejected, and checks the value round trips each time prior to that. I also tried to condense the test a bit since I was adding more code there.

The digit count is specifically the digits that round trip, so it's already the conservative value.